crates/gitbutler-oxidize/src/ext.rs

```diff
@@ -0,0 +1,123 @@
+use crate::git2_to_gix_object_id;
+use anyhow::{Context, Result};
+use gix::merge::tree::{Options, TreatAsUnresolved};
+
+pub trait GixRepositoryExt: Sized {
+    /// Configure the repository for diff operations between trees.
+    /// This means it needs an object cache relative to the amount of files in the repository.
+    fn for_tree_diffing(self) -> Result<Self>;
+
+    /// Returns `true` if the merge between `our_tree` and `their_tree` is free of conflicts.
+    /// Conflicts entail content merges with conflict markers, or anything else that doesn't merge cleanly in the tree.
+    ///
+    /// # Important
+    ///
+    /// Make sure the repository is configured [`with_object_memory()`](gix::Repository::with_object_memory()).
+    fn merges_cleanly_compat(
+        &self,
+        ancestor_tree: git2::Oid,
+        our_tree: git2::Oid,
+        their_tree: git2::Oid,
+    ) -> Result<bool>;
+
+    /// Just like the above, but with `gix` types.
+    fn merges_cleanly(
+        &self,
+        ancestor_tree: gix::ObjectId,
+        our_tree: gix::ObjectId,
+        their_tree: gix::ObjectId,
+    ) -> Result<bool>;
+
+    /// Return default label names when merging trees.
+    ///
+    /// Note that these should probably rather be branch names, but that's for another day.
+    fn default_merge_labels(&self) -> gix::merge::blob::builtin_driver::text::Labels<'static> {
+        gix::merge::blob::builtin_driver::text::Labels {
+            ancestor: Some("base".into()),
+            current: Some("ours".into()),
+            other: Some("theirs".into()),
+        }
+    }
+
+    /// Tree merge options that enforce undecidable conflicts to be forcefully resolved
+    /// to favor ours, both when dealing with content merges and with tree merges.
+    fn merge_options_force_ours(&self) -> Result<gix::merge::tree::Options>;
+
+    /// Return options suitable for merging so that the merge stops immediately after the first conflict.
+    /// It also returns the conflict kind to use when checking for unresolved conflicts.
+    fn merge_options_fail_fast(
+        &self,
+    ) -> Result<(
+        gix::merge::tree::Options,
+        gix::merge::tree::TreatAsUnresolved,
+    )>;
+
+    /// Just like [`Self::merge_options_fail_fast()`], but additionally don't perform rename tracking.
+    /// This is useful if the merge result isn't going to be used, and we are only interested in knowing
+    /// if a merge would succeed.
+    fn merge_options_no_rewrites_fail_fast(
+        &self,
+    ) -> Result<(gix::merge::tree::Options, TreatAsUnresolved)>;
+}
+
+impl GixRepositoryExt for gix::Repository {
+    fn for_tree_diffing(mut self) -> anyhow::Result<Self> {
+        let bytes = self.compute_object_cache_size_for_tree_diffs(&***self.index_or_empty()?);
+        self.object_cache_size_if_unset(bytes);
+        Ok(self)
+    }
```

Comment:
As an unrelated observation, I've seen that there are various "modes" that you can put a gix repository into. I can't help but feel like it could be a little error prone in the long run, given that the states are not represented in the types.

Reply:
I think I get the point and thought about this as well, especially in comparison to `git2::Repository` `gix::Repository` makes you think of things more, something that will obviously be forgotten.

This setting I'd call it configuration for which type-state systems would be too much, especially since I had to work to keep generics out of `gix::Repository`, something that wasn't always the case.

Something that I always thought about doing is to simply set a default object cache size, but I couldn't yet make me because it might be as surprising if the app keeps piling up memory.

So the error is on the side of loosing performance if the object cache is forgotten for some operations.

Something that I could imagine doing… very seriously so even, is that those operations that need it will automatically setup an object cache if it isn't set yet, and then remove/reset it afterwards. But… what if one wants to do a diff with minimal memory footprint and is OK for it to take longer?
`git2` shows that it's probably alright, it's just that I also don't want to take away all the viable choices.

Reply:
Sounds like you've put a lot of thought into that, and it all seems very logical. Thanks :+1:

Reply:
I wish all that thought would finally allow me to make a decision :D.
Maybe the 'auto-cache' mode is something that could be enabled by default, `git2` style, but can also be disabled as needed.
Hmmm :).